dnf/cli/commands/history.py

```diff
@@ -0,0 +1,506 @@
+                    logger.critical(
+                        _("Invalid transaction ID range definition '{}'.\n"
+                          "Use '<transaction-id>..<transaction-id>'."
+                          ).format(t))
+                    raise dnf.cli.CliError
+                cant_convert_msg = _("Can't convert '{}' to transaction ID.\n"
+                                     "Use '<number>', 'last', 'last-<number>'.")
+                try:
+                    begin_transaction_id = str2transaction_id(begin_transaction_id)
+                except ValueError:
+                    logger.critical(_(cant_convert_msg).format(begin_transaction_id))
+                    raise dnf.cli.CliError
+                try:
+                    end_transaction_id = str2transaction_id(end_transaction_id)
+                except ValueError:
+                    logger.critical(_(cant_convert_msg).format(end_transaction_id))
+                    raise dnf.cli.CliError
+                if self._require_one_transaction_id and begin_transaction_id != end_transaction_id:
+                        logger.critical(self._require_one_transaction_id_msg)
+                        raise dnf.cli.CliError
+                if begin_transaction_id > end_transaction_id:
+                    begin_transaction_id, end_transaction_id = \
+                        end_transaction_id, begin_transaction_id
+                merged_tids.add((begin_transaction_id, end_transaction_id))
+                tids.update(range(begin_transaction_id, end_transaction_id + 1))
+            else:
+                try:
+                    tids.add(str2transaction_id(t))
+                except ValueError:
+                    # not a transaction id, assume it's package name
+                    transact_ids_from_pkgname = self.output.history.search([t])
+                    if transact_ids_from_pkgname:
+                        tids.update(transact_ids_from_pkgname)
+                    else:
+                        msg = _("No transaction which manipulates package '{}' was found."
+                                ).format(t)
+                        if self._require_one_transaction_id:
+                            logger.critical(msg)
+                            raise dnf.cli.CliError
+                        else:
+                            logger.info(msg)
+
+        return sorted(tids, reverse=True), merged_tids
+
+    def _verify_toplevel_json(self):
+        fn = self.opts.transaction_filename
+
+        if "version" not in self.replay_data:
+            raise TransactionFileError(fn, _('Missing key "{key}".'.format(key="version")))
+
+        if not isinstance(self.replay_data["version"], str):
+            raise TransactionFileError(fn, _('Unexpected type of "version", string expected.'))
+
+        check_version(self.replay_data["version"], fn)
+
+    def _replay_pkg_action(self, na, action_data):
+        try:
+            na_str = "%s.%s" % (na.name, na.arch)
```

Comment:
I see a lot of splitting and formatting steps. For replay you:
1. parse NA into NEVRA object
2. Regenerate nevra from NEVRA object for To and from
3. Regenerate NA from NEVRA object

What about to reduce those steps and store full NEVRA in json?